### PR TITLE
Implement support for DLFEAT/DULBE

### PR DESCRIPTION
**Pull request by birkelund**
Hi Matias,

Please consider reviewing and pulling my branch which implements support for DLFEAT/DULBE for both the scalar and vector paths.

It also introduces checks that enforce the read/write access rules (MW_CUNITS, WS_MIN).

### CODE REVIEW

File: hw/block/nvme.c [hunk 1]
```diff
@@ -3132,8 +3130,6 @@ static int lnvm_init(NvmeCtrl *n)
         ln->lbaf.sec_mask = ((1 << ln->id_ctrl.lbaf.sec_len) - 1) <<
                                                         ln->lbaf.sec_offset;
 
-        id_ns->nuse = id_ns->ncap = id_ns->nsze = 1ULL << (ln->id_ctrl.lbaf.sec_len + ln->id_ctrl.lbaf.chk_len + ln->id_ctrl.lbaf.lun_len + ln->id_ctrl.lbaf.ch_len);
```
[the remark this answers is not in the document]
Comment (birkelund) [reply hw/block/nvme.c]: OK, I dropped that commit from the pull request.

**Comment by birkelund** [comment]
I actually have not tested with pblk, just nvme-cli and liblightnvm. But I'll look into it.

DULBE is off by default. It needs to be explicitly activated using e.g.

`nvme set-feature /dev/nvme0 -n 1 -f 0x5 -v $(( 1 << 16 ))`

**Comment by MatiasBjorling** [comment]
You're right. I had these in mind:

nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207
nvme: accessing unmapped chunk: ch:0, lun:3, chk:63, cno: 207

Which was reported on partition scan.

**Comment by birkelund** [comment]
Yeah, the hw will report that, but still return success and 0x00 to the host.

And yeah. I also seeing some corrupted reads for pblk. I'll look into it and report back.

**Comment by birkelund** [comment]
OK, I can now successfully run a fio verify test on pblk.

The issue was in function `lnvm_rw_check_chunk_read`, where I check for the read being valid with respect to `MW_CUNITS`. According to the spec a read is only valid if `READ < (WP LBA - MW_CUNITS)`, causing me to implement that check as:

```c
if (state == LNVM_CHUNK_OPEN || state == LNVM_CHUNK_CLOSED) {
    uint64_t wpp = wp < mw_cunits ? 0 : wp;
    if (state == LNVM_CHUNK_OPEN && wpp) {
        wpp = wp-mw_cunits;
    }

    if (end_sectr < wpp) {
        return 0;
    }
}
```

pblk assumes that it is allowed to read *at* the end_sectr, so the condition becomes `end_sectr <= wpp` instead. This might be something that needs to go into an errata for the spec? It was a bit unclear to me.

**Comment by MatiasBjorling** [comment]
Uh, so pblk thinks that it can read from the WP offset? It should only read up till the write pointer.

**Comment by javigon** [comment]
It has nothing to do with pblk. pblk reads WP - MW_CUNITS in the chunk as originally defined. The way it is worded in the spec creates confusion, and probably one of the formulas for the read access should be <=

**Comment by MatiasBjorling** [comment]
Here is an example:

MW_CUNITS = 16
Chunk size = 64 (0...63)

Chunk State: Open 
  WP = 32

In that case,

WP - MW_CUNITS = 16

Valid reads from 0-15. That equals to the bound checking for < 16, and not <= 16.

Maybe "Read between (WP - MW_CUNITS) LBA and WP LBA" could be reworded to make it explicit that it includes WP - MW_CUNITS.

**Comment by javigon** [comment]
That is my understanding too. It is true though that the spec is not clear about this point, as it can be interpreted 0-indexed.

We are double checking pblk, as Klaus' test look ok.

**Comment by MatiasBjorling** [comment]
Hehe, it is not the first time the zero-based value has caused questions.

**Comment by javigon** [comment]
There is a corner case in which pblk does not respect mw_cunits. I just submitted a patch to fix this upstream.

**Comment by birkelund** [comment]
I've tested @javigon's patches on my branch. pblk is happy now.

**Comment by birkelund** [comment]
Closing in favor of other.
